arrow-fx-reactor/src/main/kotlin/arrow/fx/reactor/extensions/fluxk.kt

```diff
@@ -196,79 +197,107 @@ interface FluxKConcurrent : Concurrent<ForFluxK>, FluxKAsync {
 
   override fun <A, B> CoroutineContext.racePair(fa: FluxKOf<A>, fb: FluxKOf<B>): FluxK<RacePair<ForFluxK, A, B>> =
     asScheduler().let { scheduler ->
+      val active = AtomicBooleanW(true)
+      val forward = Disposables.composite()
+
       Flux.create<RacePair<ForFluxK, A, B>> { emitter ->
         val sa = ReplayProcessor.create<A>()
         val sb = ReplayProcessor.create<B>()
 
+        emitter.onCancel(forward)
+
         val dda = fa.value()
           .subscribeOn(scheduler)
           .subscribe(sa::onNext, sa::onError)
+          .also { forward.add(it) }
+
         val ddb = fb.value()
           .subscribeOn(scheduler)
           .subscribe(sb::onNext, sb::onError)
-
-        emitter.onCancel { dda.dispose(); ddb.dispose() }
+          .also { forward.add(it) }
 
         val ffa = Fiber(sa.k(), FluxK { dda.dispose() })
         val ffb = Fiber(sb.k(), FluxK { ddb.dispose() })
 
         sa.subscribe({
-          emitter.next(RacePair.First(it, ffb))
-        }, { e -> onError(ddb, emitter, e) }, emitter::complete)
+          if (active.getAndSet(false)) {
+            emitter.next(RacePair.First(it, ffb))
+            emitter.complete()
+          }
+        }, { e -> onError(active, ddb, emitter, e) }, {})
 
         sb.subscribe({
-          emitter.next(RacePair.Second(ffa, it))
-        }, { e -> onError(dda, emitter, e) }, emitter::complete)
+          if (active.getAndSet(false)) {
+            emitter.next(RacePair.Second(ffa, it))
+            emitter.complete()
+          }
+        }, { e -> onError(active, dda, emitter, e) }, {})
       }
         .subscribeOn(scheduler)
         .k()
     }
 
   fun <A, B> onError(
+    active: AtomicBooleanW,
     other2: ReactorDisposable,
     sink: FluxSink<RacePair<ForFluxK, A, B>>,
     err: Throwable
-  ) {
+  ): Unit = if (active.getAndSet(false)) {
     other2.dispose()
     sink.error(err)
-  }
+  } else Unit
```

Comment:
huh

I mean it's the most obvious one yet I'd never used it. I'd gone for some more clever yet crappier.